pixeltable/env.py

```diff
@@ -1099,23 +1106,27 @@ def record(self, reset_exc: bool = False, **kwargs: Any) -> None:
             self.has_exc = False
             for k, v in kwargs.items():
                 if v is not None:
-                    self.resource_limits[k].update(now, *v)
+                    self.resource_limits[k].update(request_timestamp, *v)
+                    _logger.debug(f'Updated resource state: {self.resource_limits[k]}')
 
-    def record_exc(self, exc: Exception) -> None:
-        """Update self.resource_limits based on the exception headers"""
+    def record_exc(self, request_timestamp: datetime.datetime, exc: Exception) -> None:
+        """Update self.resource_limits based on the exception headers
+        Args:
+            - request_timestamp: time at which the request that caused the exception was made
+            - exc: the exception raised"""
         self.has_exc = True
 
-    def get_retry_delay(self, exc: Exception) -> float | None:
+    def get_retry_delay(self, exc: Exception, attempt: int) -> float | None:
         """Returns number of seconds to wait before retry, or None if not retryable"""
-        if len(self.resource_limits) == 0:
-            return 1.0
-        # we're looking for the maximum delay across all depleted resources
-        max_delay = 0.0
-        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        # Find the highest wait until at least 5% availability of all resources
+        max_wait = 0.0
         for limit_info in self.resource_limits.values():
-            if limit_info.remaining < 0.05 * limit_info.limit:
-                max_delay = max(max_delay, (limit_info.reset_at - now).total_seconds())
-        return max_delay if max_delay > 0 else None
+            time_until = limit_info.estimated_time_until_balance(
+                math.ceil(TARGET_RATE_LIMIT_RESOURCE_FRACT * limit_info.limit)
+            )
+            if time_until is not None:
+                max_wait = max(max_wait, time_until)
+        return max_wait if max_wait > 0 else 2.0 ** min(attempt, 4)
```

Comment:
Include some jitter in the exponential backoff?

Also maybe define constants for the "magic numbers" `2.0` and `4` instead of hardcoding them?